lib/xcknife/test_dumper.rb

```diff
@@ -186,52 +184,89 @@ def call(derived_data_folder, list_folder, extra_environment_variables = {})
 
     attr_reader :testroot
 
-    def list_tests(xctestrun, list_folder, extra_environment_variables)
-      xctestrun.reject! { |test_bundle_name, _| test_bundle_name == '__xctestrun_metadata__' }
-      xctestrun.map do |test_bundle_name, test_bundle|
+    def test_groups(xctestrun)
+      xctestrun.group_by do |test_bundle_name, _test_bundle|
         if @skip_dump_bundle_names.include?(test_bundle_name)
-          logger.info { "Skipping dumping tests in `#{test_bundle_name}` -- writing out fake event"}
-          test_specification = list_single_test(list_folder, test_bundle, test_bundle_name)
+          'single'
         elsif @naive_dump_bundle_names.include?(test_bundle_name)
-          test_specification = list_tests_with_nm(list_folder, test_bundle, test_bundle_name)
+          'nm'
         else
-          test_specification = list_tests_with_simctl(list_folder, test_bundle, test_bundle_name, extra_environment_variables)
-          wait_test_dumper_completion(test_specification.json_stream_file)
+          'simctl'
         end
+      end
+    end
+
+    # This executes naive test dumping in parallel by queueing up items onto a work queue to process
+    # with 1 new thread per processor. Results are placed onto a threadsafe spec queue to avoid writing
+    # to an object between threads, then popped off re-inserting them to our list of test results.
+    def list_tests(xctestrun, list_folder, extra_environment_variables)
+      xctestrun.reject! { |test_bundle_name, _| test_bundle_name == '__xctestrun_metadata__' }
+
+      test_runs_by_method = test_groups(xctestrun)
+      spec_queue = Queue.new
+      nm_bundle_queue = Queue.new
+      single_tests = test_runs_by_method['single'] || []
+      nm_tests = test_runs_by_method['nm'] || []
+      simctl_tests = test_runs_by_method['simctl'] || []
+
+      single_tests.each do |test_bundle_name, test_bundle|
+        logger.info { "Skipping dumping tests in `#{test_bundle_name}` -- writing out fake event" }
+        spec_queue << list_single_test(list_folder, test_bundle, test_bundle_name)
+      end
+
+      simctl_tests.each do |test_bundle_name, test_bundle|
+        test_spec = list_tests_with_simctl(list_folder, test_bundle, test_bundle_name, extra_environment_variables)
+        wait_test_dumper_completion(test_spec.json_stream_file)
 
-        test_specification
+        spec_queue << test_spec
       end
+
+      nm_tests.each { |item| nm_bundle_queue << item }
+
+      [Etc.nprocessors, nm_bundle_queue.size].min.times.map do
+        Thread.new do
+          Thread.current.abort_on_exception = true
+
+          until nm_bundle_queue.empty?
```

Comment:
actually looking at this now, I think it has the potential to race: `.empty?` returns `false` on thread 1, thread 2 pops the last item, thread 1 then waits forever in the `pop` call. sorry for not thinking of this yesterday!

Maybe it does make more sense to use a `poison` object: in first line of the `map` block, `nm_bundle_queue << :stop`, and break from the loop if `nm_bundle_queue.pop == :stop` ?